### ISSUE TITLE
Your upgrade of 113 broken all gradio apps - a fix needed ASAP this is devastating 

### ISSUE THREAD

**Issue by FurkanGozukara** [fastapi/fastapi#12133]
a very basic just gradio having app

fast api 111 fixes issue

all of my followers now reporting errors and so hard to fix all

i don't know how can you publish such a devastating bug having version please fix it ASAP

```
2024-09-06 00:12:20,515 - INFO - HTTP Request: GET https://api.gradio.app/gradio-messaging/en "HTTP/1.1 200 OK"
Running on local URL:  http://127.0.0.1:7860
2024-09-06 00:12:20,542 - INFO - HTTP Request: GET http://127.0.0.1:7860/startup-events "HTTP/1.1 200 OK"
2024-09-06 00:12:20,552 - INFO - HTTP Request: HEAD http://127.0.0.1:7860/ "HTTP/1.1 200 OK"

To create a public link, set `share=True` in `launch()`.
2024-09-06 00:12:20,673 - INFO - HTTP Request: GET https://checkip.amazonaws.com/ "HTTP/1.1 200 "
2024-09-06 00:12:21,337 - INFO - HTTP Request: GET https://api.gradio.app/pkg-version "HTTP/1.1 200 OK"
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\type_adapter.py", line 209, in __init__
    core_schema = _getattr_no_parents(type, '__pydantic_core_schema__')
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\type_adapter.py", line 98, in _getattr_no_parents
    raise AttributeError(attribute)
AttributeError: __pydantic_core_schema__

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 406, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 70, in __call__
    return await self.app(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\fastapi\applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\middleware\errors.py", line 187, in __call__
    raise exc
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\middleware\errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\gradio\route_utils.py", line 766, in __call__
    await self.simple_response(scope, receive, send, request_headers=headers)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\gradio\route_utils.py", line 782, in simple_response
    await self.app(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\middleware\exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\_exception_handler.py", line 62, in wrapped_app
    raise exc
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\_exception_handler.py", line 51, in wrapped_app
    await app(scope, receive, sender)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\routing.py", line 715, in __call__
    await self.middleware_stack(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\routing.py", line 735, in app
    await route.handle(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\routing.py", line 288, in handle
    await self.app(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\routing.py", line 76, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\_exception_handler.py", line 62, in wrapped_app
    raise exc
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\_exception_handler.py", line 51, in wrapped_app
    await app(scope, receive, sender)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\starlette\routing.py", line 73, in app
    response = await f(request)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\fastapi\routing.py", line 291, in app
    solved_result = await solve_dependencies(
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\fastapi\dependencies\utils.py", line 639, in solve_dependencies
    ) = await request_body_to_args(  # body_params checked above
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\fastapi\dependencies\utils.py", line 810, in request_body_to_args
    fields_to_extract = get_model_fields(first_field.type_)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\fastapi\_compat.py", line 283, in get_model_fields
    return [
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\fastapi\_compat.py", line 284, in <listcomp>
    ModelField(field_info=field_info, name=name)
  File "<string>", line 6, in __init__
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\fastapi\_compat.py", line 109, in __post_init__
    self._type_adapter: TypeAdapter[Any] = TypeAdapter(
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\type_adapter.py", line 211, in __init__
    core_schema = _get_schema(type, config_wrapper, parent_depth=_parent_depth + 1)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\type_adapter.py", line 81, in _get_schema
    schema = gen.generate_schema(type_)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 490, in generate_schema
    schema = self._generate_schema(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 721, in _generate_schema
    schema = self._generate_schema_inner(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 727, in _generate_schema_inner
    return self._annotated_schema(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 1697, in _annotated_schema
    schema = self._apply_annotations(source_type, annotations)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 1765, in _apply_annotations
    schema = get_inner_schema(source_type)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_schema_generation_shared.py", line 82, in __call__
    schema = self._handler(__source_type)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 1847, in new_handler
    schema = metadata_get_schema(source, get_inner_schema)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 1843, in <lambda>
    lambda source, handler: handler(source)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_schema_generation_shared.py", line 82, in __call__
    schema = self._handler(__source_type)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 1746, in inner_handler
    schema = self._generate_schema(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 721, in _generate_schema
    schema = self._generate_schema_inner(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 747, in _generate_schema_inner
    return self.match_type(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 830, in match_type
    return self._match_generic_type(obj, origin)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 854, in _match_generic_type
    return self._union_schema(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 1141, in _union_schema
    choices.append(self.generate_schema(arg))
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 490, in generate_schema
    schema = self._generate_schema(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 721, in _generate_schema
    schema = self._generate_schema_inner(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 747, in _generate_schema_inner
    return self.match_type(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 834, in match_type
    return self._unknown_type_schema(obj)
  File "R:\Image_Caption_Editor_v6\venv\lib\site-packages\pydantic\_internal\_generate_schema.py", line 393, in _unknown_type_schema
    raise PydanticSchemaGenerationError(
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'starlette.requests.Request'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.6/u/schema-for-unknown-type

```

**Comment by drikster80** [comment]
It looks like this latest version broke the vllm project as well:
https://github.com/vllm-project/vllm/issues/8212

Not quite sure how to fix it so just pinning `fastapi == 0.112.2` for now. :-(

**Comment by FurkanGozukara** [comment]
> It looks like this latest version broke the vllm project as well: [vllm-project/vllm#8212](https://github.com/vllm-project/vllm/issues/8212)
> 
> Not quite sure how to fix it so just pinning `fastapi == 0.112.2` for now. :-(

it broken literally all of the gradio apps. it is such a devastating bug. i have 10s of apps and updating installers and making all fast api lower version is so problematic

especially many apps are remote repos that i use from other people

**Comment by cobanov** [comment]
https://github.com/bmaltais/kohya_ss/issues/2783#issuecomment-2332444719

**Comment by tiangolo** [comment]
@FurkanGozukara please share a self-contained [minimal self-contained example](https://stackoverflow.com/help/minimal-reproducible-example) that replicates the issue so that I can copy-paste it, run it, and see your error. I just tried with the getting-started Gradio app, and it all seems to work.

Now, the right place to start would have been a GitHub Discussion, filling the form, that would have guided you to provide a reproducible example.

And then, please be kind, your post is aggressive and demeaning, this is open source, you get it for free, you are not paying nor contributing anything here, so at least, be kind.

**Comment by neoyxm** [comment]
same issue in my gradio project. 

rollback to 0.112.2 as the workaround

**Comment by neoyxm** [comment]
> @FurkanGozukara please share a self-contained [minimal self-contained example](https://stackoverflow.com/help/minimal-reproducible-example) that replicates the issue so that I can copy-paste it, run it, and see your error. I just tried with the getting-started Gradio app, and it all seems to work.
> 
> Now, the right place to start would have been a GitHub Discussion, filling the form, that would have guided you to provide a reproducible example.
> 
> And then, please be kind, your post is aggressive and demeaning, this is open source, you get it for free, you are not paying nor contributing anything here, so at least, be kind.

0.113.0's release pre-commit failed. 

Is it the root cause?

<img width="830" alt="截屏2024-09-06 16 51 55" src="https://github.com/user-attachments/assets/9805e0c2-3d21-48a4-a0c4-b64b8c8e7bfc">

**Comment by FurkanGozukara** [comment]
> @FurkanGozukara please share a self-contained [minimal self-contained example](https://stackoverflow.com/help/minimal-reproducible-example) that replicates the issue so that I can copy-paste it, run it, and see your error. I just tried with the getting-started Gradio app, and it all seems to work.
> 
> Now, the right place to start would have been a GitHub Discussion, filling the form, that would have guided you to provide a reproducible example.
> 
> And then, please be kind, your post is aggressive and demeaning, this is open source, you get it for free, you are not paying nor contributing anything here, so at least, be kind.

sorry for my aggressiveness i was really tired and frustrated from all the bugs yesterday

and your commit literally breaks everything here a minimal app tested on gradio 4.42.0

open app and enter a folder path thats it

```
import gradio as gr
import os

def scan_folders(folder_path):
    if not os.path.isdir(folder_path):
        return "Error: The provided path is not a valid directory."
    
    result = []
    for root, dirs, files in os.walk(folder_path):
        for dir in dirs:
            result.append(os.path.join(root, dir))
    
    if not result:
        return "No subfolders found in the specified directory."
    
    return "\n".join(result)

iface = gr.Interface(
    fn=scan_folders,
    inputs=gr.Textbox(label="Enter folder path"),
    outputs=gr.Textbox(label="Scanned folders"),
    title="Folder Scanner",
    description="Enter a folder path to scan all subfolders within it."
)

iface.launch()
```

**Comment by tiangolo** [comment]
Thanks for the minimal app! I'm being able to make it break with that, so now I can debug it.

---

I guess Gradio and vllm are using some of the FastAPI internals and/or monkey patching them, which is not really supported/documented, but I'll see how to fix it. :nerd_face:

I also need to see what they need and why they are monkeypatching or using internals, to provide some other and official way to do it.

**Comment by FurkanGozukara** [comment]
> Thanks for the minimal app! I'm being able to make it break with that, so now I can debug it.
> 
> I guess Gradio and vllm are using some of the FastAPI internals and/or monkey patching them, which is not really supported/documented, but I'll see how to fix it. 🤓
> 
> I also need to see what they need and why they are monkeypatching or using internals, to provide some other and official way to do it.

I just pinged your comment they may help

https://github.com/gradio-app/gradio/issues/9278#issuecomment-2333675972

**Comment by jhj0517** [comment]
For me updating the gradio version fixed the bug,
```
pip install -U gradio==4.43.0
```

The bug is reproducible with:

```
gradio==4.29.0
fastapi==0.113.0 
```

Works well with:
```
gradio==4.43.0
fastapi==0.112.4
```

The fastapi versions are bound to gradio, which are installed automatically when you install gradio.

**Comment by peeter2** [comment]
there is no such thing as version 4.43.0 on gradio changelog page.
the last version they have is 4.42.0.

**Comment by jhj0517** [comment]
@peeter2 The log may not be updated yet. 4.43.0 is released on pypi.
https://pypi.org/project/gradio/#history

**Comment by tiangolo** [comment]
About Gradio, I debugged it and reported here: https://github.com/gradio-app/gradio/issues/9278#issuecomment-2333875064. It's something that should be refactored in Gradio (the current code is doing something that shouldn't be done), they are aware of it and working on the fix.

Now let's check vllm. :nerd_face:

**Comment by tiangolo** [comment]
@drikster80 can you provide a self-contained [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of how vllm is not working?

I'm trying it here with the quickstart and it all seems to work.

**Comment by tiangolo** [comment]
@cobanov about https://github.com/bmaltais/kohya_ss/issues/2783, that repo seems to be a Gradio app, so that would be fixed once they fix and release that in Gradio.

**Comment by drikster80** [comment]
> @drikster80 can you provide a self-contained [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of how vllm is not working?
> 
> I'm trying it here with the quickstart and it all seems to work.

Hi @tiangolo. I'll have some time to dig into this in about 3 hours. Thanks for your support. Installing and starting the software seems to work fine. It only occurs when you attempt to make an inference call through the openai-api, which I'm not quite sure how to debug at the moment (and I'm not terribly familiar with that part of the vllm codebase). Will dig deeper shortly.

**Comment by FurkanGozukara** [comment]
all apps that uses specific gradio before to 4.43.0 broken right at this moment :/

**Comment by tiangolo** [comment]
@drikster80 interesting, I wouldn't think those lines would be a problem as those are parameters for functions, not Pydantic models used for inputs in an API. :thinking:

**Comment by joerunde** [comment]
> all apps that uses specific gradio before to 4.43.0 broken right at this moment :/

@FurkanGozukara There's no need to be dramatic. Installing dependencies is the responsibility of the application, not the libraries it consumes. If you wrote an app and didn't ensure it has reproducible builds, that's on you

**Comment by patrickblackjr** [comment]
> > > all apps that uses specific gradio before to 4.43.0 broken right at this moment :/
> > 
> > 
> > @FurkanGozukara There's no need to be dramatic. Installing dependencies is the responsibility of the application, not the libraries it consumes. If you wrote an app and didn't ensure it has reproducible builds, that's on you
> 
> I don't know how many open source ai apps you used but there are countless amazing apps made by researchers and not maintained anymore or rarely maintained
> 
> So becuase of this reason I have to manually fix my installers for 10s of such apps and re-upload them and if I miss it will be broken next time a user uses it
> 
> I have a c# background I hate python because of this. 0 backwards. With Microsoft I can load 12 years old project and it will compile straight in latest Microsoft visual studio and work

Would your tests not have caught this? I have tests that run for dependency updates to ensure core functionality doesn't break if I upgrade a version for a dependency. 

I don't think this is a Python-specific problem, but more a community mindset difference. C# is often used in larger enterprises, which typically are not going to be "bleeding edge" or do frequent version upgrades, so the general mindset is going to be prioritize backwards compatibility over new features.

**Comment by drikster80** [comment]
@tiangolo, I updated the vllm issue with the following confirmed fix. Any idea why this works in 0.112.2 by not >=0.113.0??

**Update to vLLM Issue:**
I believe I was able to find a solution to this. It is related to https://github.com/openai/openai-python/issues/1454 and https://github.com/fastapi/fastapi/discussions/11952

Not sure why it works with fastapi 0.112.2 but fails in 0.113.0

Problem line:

https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/api_server.py#L286

```python
async def create_chat_completion(request: ChatCompletionRequest,
                                 raw_request: Request):
```
Confirmed Fix:

```python
async def create_chat_completion(request: Annotated[dict, ChatCompletionRequest],
                                 raw_request: Request):
```

**Comment by joerunde** [comment]
> I don't know how many open source ai apps you used but there are countless amazing apps made by researchers and not maintained anymore or rarely maintained

> So becuase of this reason I have to manually fix my installers for 10s of such apps and re-upload them and if I miss it will be broken next time a user uses it

Sure, there are lots of good projects out there. And if they're not maintained and you took on the burden of maintenance for your users, then there are steps you can take to make installs reproducible. Try poetry lockfiles.

Yelling at maintainers of projects that you use isn't productive. Did you just expect the alternative that gradio would pin all dependencies to specific versions so that nothing could ever break? Then you'd have to bug them for a new release every time you need a new version of a dependency 🙃 

> I have a c# background I hate python because of this. 0 backwards. With Microsoft I can load 12 years old project and it will compile straight in latest Microsoft visual studio and work

If you use a lockfile, then you can come back to your python project in 12 years and run it just fine 😉

**Comment by tiangolo** [comment]
@drikster80 

> I believe I was able to find a solution to this. It is related to https://github.com/openai/openai-python/issues/1454 and https://github.com/fastapi/fastapi/discussions/11952

I would think it's not related, those two links are asking for support for TypedDict for query params, that's different from body requests.

> Problem line:

> [vllm-project/vllm@main/vllm/entrypoints/openai/api_server.py#L286](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/api_server.py?rgh-link-date=2024-09-06T21%3A03%3A29Z#L286)

This line seems to declare a `ChatCompletionRequest`, that inherits from `OpenAIBaseModel` which inherits from Pydantic's `BaseModel`, so, I would think it's not related to the `TypedDict` issue.

---

Maybe you could provide me with steps to reproduce the problem. I'm not a user of vllm myself, I tried the basic stuff in the quick start and it all worked, maybe if you tell me what are the steps, keys I need to get and set from OpenAI, the command to run, and request to make to trigger the error, I could then debug it. :nerd_face:

**Comment by drikster80** [comment]
@tiangolo, thanks for your patience and support. Here is the most simple way to recreate the issue

## Triggering the issue from fastapi upgrade:

**Pre-requisites:**
- Requires system with GPU and ability to pass GPU into docker container (e.g. nvidia-container-toolkit)

Download and start the latest vllm container:
```
docker run --gpus all -it --rm --network=host --ipc=host --entrypoint /bin/bash vllm/vllm-openai:latest
```

### Working Example with 0.112.2:

Show current fastapi version:
```
python3 -c "import fastapi; print(fastapi.__version__)"
0.112.2
```

Start server with small model

```
python3 -m vllm.entrypoints.openai.api_server --model facebook/opt-125m
```

POST to 'v1/chat/completions'

```
curl http://localhost:8000/v1/chat/completions \
-H "Content-Type: application/json" \
-d '{
"model": "facebook/opt-125m",
"messages": [
{"role": "system", "content": "You are a helpful assistant."},
{"role": "user", "content": "Who won the world series in 2020?"}
]
}'
```

### Non-working example after upgrading fastapi:

Upgrade fastapi to 0.113.0 or higher
```
pip install --upgrade fastapi==0.113.0
```

Start openai-compatible api_server:
```
python3 -m vllm.entrypoints.openai.api_server --model facebook/opt-125m
```

From outside the container, attempt POST to 'v1/chat/completions':

```
curl http://localhost:8000/v1/chat/completions \
-H "Content-Type: application/json" \
-d '{
"model": "facebook/opt-125m",
"messages": [
{"role": "system", "content": "You are a helpful assistant."},
{"role": "user", "content": "Who won the world series in 2020?"}
]
}'
```


### vLLM Container Dev setup using pre-compiled C-binaries (saves hours of compiling):

Install devel packages for Nvidia
```
apt-get update && apt-get install -y --no-install-recommends libtinfo5 libncursesw5 \
	cuda-cudart-dev-12-4=12.4.127-1 \
	cuda-command-line-tools-12-4=12.4.1-1 \
	cuda-minimal-build-12-4=12.4.1-1 \
	cuda-libraries-dev-12-4=12.4.1-1 \
	cuda-nvml-dev-12-4=12.4.127-1 \
	cuda-nvprof-12-4=12.4.127-1 \
	libnpp-dev-12-4=12.2.5.30-1 \
	libcusparse-dev-12-4=12.3.1.170-1 \
	libcublas-dev-12-4=12.4.5.8-1 \
	libnccl2=2.21.5-1+cuda12.4 \
	libnccl-dev=2.21.5-1+cuda12.4 \
	cuda-nsight-compute-12-4=12.4.1-1
```

### Build vllm editable using precompiled binaries
```
cd /vllm
cp /usr/local/lib/python3.10/dist-packages/vllm/*.so /vllm/vllm
VLLM_USE_PRECOMPILED=1 pip install -e .
```

Run API server
```
python3 ./vllm/entrypoints/openai/api_server.py --model facebook/opt-125m
```

Example of an inference request:
```
curl http://localhost:8000/v1/chat/completions \
-H "Content-Type: application/json" \
-d '{
  "model": "facebook/opt-125m",
  "messages": [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Who won the world series in 2020?"}
  ],
  "chat_template": "{% if messages[0][\"role\"] == \"system\" %}{{ messages[0][\"content\"] }}\n{% endif %}{% for message in messages[1:] %}{% if message[\"role\"] == \"user\" %}Human: {{ message[\"content\"] }}\n{% elif message[\"role\"] == \"assistant\" %}Assistant: {{ message[\"content\"] }}\n{% endif %}{% endfor %}Assistant:",
  "max_tokens": 100
}'
```

**Comment by pachewise** [comment]
@tiangolo @drikster80 posted thoughts and a smaller reproducible example for the vllm case here: https://github.com/vllm-project/vllm/issues/8212#issuecomment-2339777580

imo there is still a class of "0.113 exposed broken pydantic schema" issues that may need to be addressed at the fastapi layer, but the vllm issue seems different from the gradio one.

**Comment by dzhulgakov** [comment]
@tiangolo - I also see a serious performance regression caused by the same change in `0.113.0`: the request parsing now takes much longer because FastAPI spends a lot of time constructing the schema on every request. For our serving CPU utilization more than doubled after upgrading to `0.113.0` from `0.112.4`. The stack trace from profiler involves `get_model_fields` that was added here:

https://github.com/fastapi/fastapi/blob/0.113.0/fastapi/dependencies/utils.py#L810

<img width="1916" alt="image" src="https://github.com/user-attachments/assets/70cbcdd5-66f6-46eb-95c2-2c3607a19827">

I've confirmed that `0.114.0` doesn't fix the issue. I'll try to come up with a minimal repro, but perhaps you may have some idea of what's going on.

**Comment by tiangolo** [comment]
Thanks @drikster80 and @pachewise!

Checking @pachewise's code in that comment, I was able to reduce the error reproduction to only using Pydantic's `TypeAdapter`. I commented with it here: https://github.com/vllm-project/vllm/issues/8212#issuecomment-2342848855

It seems it's fixed by upgrading Pydantic to 2.9.0.

<details>
<summary>
Code example
</summary>

```Python
from typing_extensions import Annotated
from typing import List
from vllm.entrypoints.chat_utils import (
    ChatCompletionMessageParam,
)
from vllm.entrypoints.openai.protocol import ChatCompletionRequest


from pydantic import TypeAdapter


for name, field in ChatCompletionRequest.model_fields.items():
    print(name, field)
    TypeAdapter(Annotated[List[ChatCompletionMessageParam], field])
```

</details>

It wasn't breaking in FastAPI before because the logic before 0.113.0 wasn't using `TypeAdapter` yet in that part of the code, and it seems that in the previous version of Pydantic there was a bug in it (not sure where, but it's already solved in 2.9.0).

---

@drikster80 could you check if upgrading Pydantic solves it for you? I currently don't have nvidia-container-toolkit and I know setting up/debugging CUDA will take quite a bit of time. :sweat_smile:

**Comment by tiangolo** [comment]
Thanks @dzhulgakov, I'll look into it.

**Comment by tiangolo** [comment]
@dzhulgakov thanks for the detailed report, the flamegraph was very useful.

It should be fixed with https://github.com/fastapi/fastapi/pull/12184, available in FastAPI 0.114.1, just released. :rocket:

**Comment by tiangolo** [comment]
About the original Gradio issue, it seems to have been solved here: https://github.com/gradio-app/gradio/pull/9279

**Comment by dzhulgakov** [comment]
@tiangolo - thanks! `0.114.1` indeed fixes our perf degradation

**Comment by drikster80** [comment]
@tiangolo, confirmed that upgrading to pydantic >2.9.0 fixes the vllm issue! I think this issue can be finally closed out. 👍

Thanks for all your help.

**Comment by pachewise** [comment]
@tiangolo found the issue related to the vllm errors: https://github.com/fastapi/fastapi/pull/12197

**Comment by tiangolo** [comment]
Thanks for the fix @pachewise! :rocket: 

It will be available in FastAPI `0.115.1` in the next hours. :tada:
